src/test/java/engineering/swat/watch/RecursiveWatchTests.java

```diff
@@ -0,0 +1,120 @@
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+import engineering.swat.watch.WatchEvent.Kind;
+
+class RecursiveWatchTests {
+    private final Logger logger = LogManager.getLogger();
+
+    private TestDirectory testDir;
+
+    @BeforeEach
+    void setup() throws IOException {
+        testDir = new TestDirectory();
+    }
+
+    @AfterEach
+    void cleanup() throws IOException {
+        if (testDir != null) {
+            testDir.close();
+        }
+    }
+
+    @BeforeAll
+    static void setupEverything() throws IOException {
+        Awaitility.setDefaultTimeout(TestHelper.NORMAL_WAIT);
+    }
+
+
+    @Test
+    void newDirectoryWithFilesChangesDetected() throws IOException {
+        var target = new AtomicReference<Path>();
+        var created = new AtomicBoolean(false);
+        var changed = new AtomicBoolean(false);
+        var watchConfig = Watcher.watch(testDir.getTestDirectory(), WatchScope.PATH_AND_ALL_DESCENDANTS)
+            .onEvent(ev -> {
+                    logger.debug("Event received: {}", ev);
+                    if (ev.calculateFullPath().equals(target.get())) {
+                        switch (ev.getKind()) {
+                            case CREATED:
+                                created.set(true);
+                                break;
+                            case MODIFIED:
+                                changed.set(true);
+                                break;
+                            default:
+                                break;
+                        }
+                    }
+            });
+
+        try (var activeWatch = watchConfig.start() ) {
+            var freshFile = Files.createTempDirectory(testDir.getTestDirectory(), "new-dir").resolve("test-file.txt");
+            target.set(freshFile);
+            logger.debug("Interested in: {}", freshFile);
+            Files.writeString(freshFile, "Hello world");
+            await("New files should have been seen").untilTrue(created);
+            Files.writeString(freshFile, "Hello world 2");
+            await("Fresh file change have been detected").untilTrue(changed);
+        }
+    }
+
+    @Test
+    void correctRelativePathIsReported() throws IOException {
+        Path relative = Path.of("a","b", "c", "d.txt");
+        var seen = new AtomicBoolean(false);
+        var watcher = Watcher.watch(testDir.getTestDirectory(), WatchScope.PATH_AND_ALL_DESCENDANTS)
+            .onEvent(ev -> {
+                logger.debug("Seen event: {}", ev);
+                if (ev.getRelativePath().equals(relative)) {
+                    seen.set(true);
+                }
+            });
+
+        try (var w = watcher.start()) {
+            var targetFile = testDir.getTestDirectory().resolve(relative);
+            Files.createDirectories(targetFile.getParent());
+            Files.writeString(targetFile, "Hello World");
+            await("Nested path is seen").untilTrue(seen);
+        }
```

Comment:
What about creating a new directory inside the new directory? (And then a new file inside the nested-nested directory)

Reply:
Nm, this is already what's happening in the current test